client_authentication.go

```diff
@@ -251,6 +262,43 @@ func (f *Fosite) DefaultClientAuthenticationStrategy(ctx context.Context, r *htt
 	return client, nil
 }
 
+// Tries to read client ID from (yet) unverified token, if request came unauthenticated
+func jwtBearerNoClientAuthRequest(r *http.Request, form url.Values, originalErr error) (clientID, clientSecret string, err error) {
+	if grantType := form.Get("grant_type"); grantType != grantTypeJWTBearer {
+		return "", "", originalErr
+	}
+	assertion := form.Get("assertion")
+
+	if assertion == "" {
+		return "", "", errorsx.WithStack(ErrInvalidRequest.WithHint("Required parameter assertion is missing."))
+	}
+	// do not validate the key (hence noop token func), all this is just to get the issuer as client ID
+	token, err := josejwt.ParseSigned(assertion)
+	// if there is no assertion is not a jwt (signature ignored), then authentication of the client has obviously failed
+	if err != nil {
+		return "", "", errorsx.WithStack(ErrInvalidRequest.WithHint("JWT in assertion can't be parsed."))
+	}
+
+	// parse claims without looking at token
+	claims := jwt.MapClaims{}
```

Comment:
If you use a struct here you can save all the type assertion code later:

```suggestion
	claims := struct{Issuer string `json:"issuer"`}{}
```